modules/standard/IO.chpl

```diff
@@ -2282,6 +2310,177 @@ The :record:`fileWriter` type supports 3 fields:
+    } else if t == ioLiteral {
+      writer._writeLiteral(x.val);
+    } else if t == ioNewline || t == ioChar {
+      writer._writeOne(writer.kind, x, writer.getLocaleOfIoRequest());
+    } else if isClassType(t) || isAnyCPtr(t) {
+      _encodeClassOrPtr(writer, x);
+    } else if isUnionType(t) {
+      _encodeUnion(writer, x);
+    } else {
+      x.encodeTo(writer.withFormatter(new DefaultWriter()));
+    }
+  }
+
+  //
+  // Writes a field name, followed by the literal string " = ", then writes
+  // the value of 'x'.
+  //
+  proc writeField(writer:fileWriter, name: string, const x: ?) : void throws {
+    if !firstField then
+      writer._writeLiteral(", ");
+
+    if !name.isEmpty() {
+      writer._writeLiteral(name);
+      writer._writeLiteral(" = ");
+    }
+
+    writer.write(x);
+    firstField = false;
+  }
+
+  proc writeTypeStart(w: fileWriter, type T) throws {
+
+    // TODO: Arrays, and array-like things (e.g. lists)
+    if _inheritLevel == 0 {
+      if isClassType(T) then
+        w._writeLiteral("{");
+      else if isRecordType(T) || isTupleType(T) then
+        w._writeLiteral("(");
+      else
+        throw new Error("unhandled type in writeTypeStart");
+    }
+
+    _inheritLevel += 1;
+  }
+
+  proc writeTypeEnd(w: fileWriter, type T) throws {
+
+    if _inheritLevel == 1 {
+      if isClassType(T) then
+        w._writeLiteral("}");
+      else if isRecordType(T) then
+        w._writeLiteral(")");
+      else if isTupleType(T) {
+        if T.size == 1 then w._writeLiteral(",)");
+        else w._writeLiteral(")");
+      } else
+        throw new Error("unhandled type in writeTypeEnd");
+    }
+
+    _inheritLevel -= 1;
+  }
+}
+
+
+record DefaultReader {
+  var firstField = true;
+  var _inheritLevel = 0;
+  var _readStart, _readEnd : bool;
+
+  proc decodeClass(reader:fileReader, type readType) : readType throws {
+    if isGenericType(readType) then compilerError("can't read generic class '" + readType:string + "'");
+
+    if isNilableClassType(readType) {
+      if reader.matchLiteral("nil") {
+        return nil:readType;
+      }
+    }
+
+    return new readType(reader.withFormatter(new DefaultReader()));
+  }
+
+  proc decode(reader:fileReader, type readType) : readType throws {
+    if isNumericType(readType) || isBoolType(readType) || isEnumType(readType) ||
+       readType == string || readType == bytes {
+      var x : readType;
+      reader._readOne(reader.kind, x, here);
+      return x;
+    } else if isClassType(readType) {
+      return decodeClass(reader, readType);
+    } else if canResolveTypeMethod(readType, "decodeFrom", reader) {
+      return readType.decodeFrom(reader.withFormatter(new DefaultReader()));
+    } else {
+      return new readType(reader.withFormatter(new DefaultReader()));
+    }
+  }
+
+  proc readField(r:fileReader, key: string, type T) throws {
+    return decodeField(r, key, T);
+  }
+
+  proc decodeField(r:fileReader, key: string, type T) throws {
```

Comment:
Why isn't this  function just called `readField`?

Reply:
Good point! It must have slipped past me during some cleanup.